src/topology/continuous_function/stone_weierstrass.lean

```diff
@@ -0,0 +1,350 @@
+  and finally using compactness to produce the desired function `h`
+  as a maximum over finitely many `x` of a minimum over finitely many `y` of the `g x y`.
+  -/
+  dsimp [set.separates_points_strongly] at sep,
+
+  let g : Π x y, L := λ x y, (sep f x y).some,
+  have w₁ : ∀ x y, (g x y : C(X, ℝ)) x = f x := λ x y, (sep f x y).some_spec.1,
+  have w₂ : ∀ x y, g x y y = f y := λ x y, (sep f x y).some_spec.2,
+
+  -- For each `x y`, we define `U x y` to be `{z | f z - ε < g x y z}`,
+  -- and observe this is a neighbourhood of `y`.
+  let U : Π x y, set X := λ x y, {z | f z - ε < g x y z},
+  have U_nhd_y : ∀ x y, U x y ∈ 𝓝 y,
+  { intros x y,
+    refine mem_nhds_sets _ _,
+    { rw [show U x y = (f - g x y : C(X, ℝ)) ⁻¹' set.Iio ε, { ext z, simp [sub_lt], }],
+      exact is_open.preimage (by continuity) is_open_Iio, },
+    { rw [set.mem_set_of_eq, w₂],
+      exact sub_lt_self _ pos, }, },
+
+  -- Fixing `x` for a moment, we have a family of functions `λ y, g x y`
+  -- which on different patches (the `U x y`) are greater than `f z - ε`.
+  -- Taking the supremum of these functions
+  -- indexed by a finite collection of patches which cover `X`
+  -- will give us an element of `A` that is globally greater than `f z - ε`
+  -- and still equal to `f x` at `x`.
+
+  -- Since `X` is compact, for every `x` there is some finset `ys t`
+  -- so the union of the `U x y` for `y ∈ ys x` still covers everything.
+  let ys : Π x, finset X := λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some,
+  let ys_w : ∀ x, (⋃ y ∈ ys x, U x y) = ⊤ :=
+    λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some_spec,
+
+  have ys_nonempty : ∀ x, (ys x).nonempty :=
+    λ x, set.nonempty_of_union_eq_top_of_nonempty _ _ nX (ys_w x),
+
+  -- Thus for each `x` we have the desired `h x : A` so `f z - ε < h x z` everywhere
+  -- and `h x x = f x`.
+  let h : Π x, L := λ x,
+    ⟨(ys x).sup' (ys_nonempty x) (λ y, (g x y : C(X, ℝ))),
+      finset.sup'_mem _ sup_mem _ _ _ (λ y _, (g x y).2)⟩,
+  have lt_h : ∀ x z, f z - ε < h x z,
+  { intros x z,
+    obtain ⟨y, ym, zm⟩ := set.exists_set_mem_of_union_eq_top _ _ (ys_w x) z,
+    dsimp [h],
+    simp only [finset.lt_sup'_iff, continuous_map.sup'_apply],
+    exact ⟨y, ym, zm⟩, },
+  have h_eq : ∀ x, h x x = f x, by { intro x, simp [w₁], },
+
+  -- For each `x`, we can take the finite intersection of the `V x y` corresponding to `y ∈ ys x`.
+  let W : Π x, set X := λ x, {z | h x z < f z + ε},
+  have W_preimage : ∀ x, W x = (h x - f : C(X, ℝ)) ⁻¹' set.Iio ε,
+  { intro x, ext z,
+    dsimp only [W, set.mem_set_of_eq],
+    rw [←sub_lt_iff_lt_add', ←pi.sub_apply],
+    apply iff.refl, },
+  -- This is still a neighbourhood of `x`.
+  have W_nhd : ∀ x, W x ∈ 𝓝 x,
+  { intros x,
+    refine mem_nhds_sets _ _,
+    { rw W_preimage,
+      exact is_open.preimage (by continuity) is_open_Iio, },
+    { dsimp only [W, set.mem_set_of_eq],
+      rw h_eq,
+      exact lt_add_of_pos_right _ pos}, },
```

Comment:
@sgouezel, this seems incredibly cumbersome, just to show that `W x` is open. Is there existing API for this? I couldn't find it.

Reply:
Yes, you were missing `is_open_lt`.

Reply:
Great. I've added a helper lemma so `continuity` can finish after `is_open_lt`.